glide-core/src/client/value_conversion.rs

```diff
@@ -607,6 +610,286 @@ pub(crate) fn convert_to_expected_type(
+        Using `XInfoStreamReturnType` recursively for maps' value type.
+        */
+        ExpectedReturnType::XInfoStreamReturnType => {
+            dbg!(value.clone());
+            match value {
+                // a RESP3 response
+                Value::Map(map) => {
+                    dbg!("map");
+                    let result = map
+                    .into_iter()
+                    .map(|(key, inner_value)| {
+                        dbg!(key.clone());
+                        dbg!(inner_value.clone());
+
+                        let converted_key = convert_to_expected_type(key, Some(ExpectedReturnType::SimpleString))?;
+                        if converted_key == Value::SimpleString("groups".into())
+                        || converted_key == Value::SimpleString("consumers".into()) {
+                            let Value::Array(nested_array) = inner_value.clone() else {
+                                // "groups" could mapped to an integer if command submitted without `FULL` keyword
+                                if matches!(inner_value.clone(), Value::Int(_)) {
+                                    return Ok((converted_key, inner_value));
+                                }
+                                return Err((ErrorKind::TypeError, "Incorrect value type received /////").into());
+                            };
+                            // array is empty or already converted (a RESP3 response) - do nothing
+                            if nested_array.is_empty() || matches!(nested_array[0], Value::Map(_)) {
+                                dbg!("already converted (a RESP3 response) - do nothing");
+                                Ok((converted_key, inner_value))
+                            } else {
+                                dbg!("else");
+                                //////////////
+                                let Value::Array(nested_array) = inner_value.clone() else {
+                                    return Err((ErrorKind::TypeError, "Incorrect value type received ----").into());
+                                };
+                                // nested array is a 2D array actually
+                                let converted_array: Vec<Value> = nested_array.into_iter().map(|elem| {
+                                    let Value::Array(mut nested_sub_array) = elem else {
+                                        return Err((ErrorKind::TypeError, "Incorrect value type received +++").into());
+                                    };
+                                    let mut converted = Vec::with_capacity(nested_sub_array.len() / 2);
+                                    // convert it to a map, while converting keys to `SimpleString`s
+                                    while !nested_sub_array.is_empty() {
+                                        let submap_key = convert_to_expected_type(nested_sub_array.remove(0), Some(ExpectedReturnType::SimpleString))?;
```

Comment:
I think array.remove(0) may be worse performance-wise than array[0].clone() here. Since we're removing the first element, everything element after this has to get shifted, which will add time proportional to how long the array is. Additionally, I think every element that is shifted gets copied to perform the shift, which adds time proportional to how complex the element is.